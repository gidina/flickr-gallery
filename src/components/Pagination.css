.center {
  text-align: center;
}

.pagination {
  list-style: none;
  display: inline-block;
}

.pagination li {
  cursor: pointer;
  color: black;
  float: left;
  padding: 8px 16px;
  text-decoration: none;
  transition: background-color 0.3s;
  border: 1px solid #ddd;
  margin: 0 4px;
}

.pagination li.active {
  background-color: #f55;
  color: white;
  border: 1px solid #f55;
}

.pagination li:hover:not(.active) {
  background-color: #ddd;
}
