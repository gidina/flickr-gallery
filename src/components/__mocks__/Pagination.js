import React from "react";
export default class Pagination extends React.Component {
    render() {
        return <pagination-mock {...this.props}>{this.props.children}</pagination-mock>;
    }
}