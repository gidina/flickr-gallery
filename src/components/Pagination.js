import React, { Component, Fragment } from "react";
import "./Pagination.css";

const defaultActive = 0;
const numItemsPage = 8;

class Pagination extends Component {
  state = { activePage: defaultActive };
  componentDidMount = () => {
    this.props.onPageChange(0, numItemsPage - 1);
  };
  onPageChangeHandler = indexPage => {
    const startIndexItem = indexPage * numItemsPage;
    const endIndexItem = startIndexItem + (numItemsPage - 1);

    this.props.onPageChange(startIndexItem, endIndexItem);
    this.setState({ activePage: indexPage });
  };
  render = () => {
    const { activePage } = this.state;
    const { numItems } = this.props;
    const totalPages = numItems ? Math.trunc(numItems / numItemsPage) + 1 : 0;
    const pagines = Array.from(Array(totalPages).keys());

    return (
      <div className="center">
        <ul className="pagination">
          <li
            onClick={() =>
              activePage > 0 ? this.onPageChangeHandler(activePage - 1) : null
            }
          >
            &laquo;
          </li>
          {pagines.map(indexPage => (
            <li
              key={indexPage}
              className={indexPage === activePage ? "active" : null}
              onClick={() => this.onPageChangeHandler(indexPage)}
            >
              {indexPage + 1}
            </li>
          ))}
          <li
            onClick={() =>
              activePage < totalPages - 1
                ? this.onPageChangeHandler(activePage + 1)
                : null
            }
          >
            &raquo;
          </li>
        </ul>
      </div>
    );
  };
}

// Pagination.protoTypes = {
// Â numItems: PropTypes.number.integer.isRequired
// onPageChange: PropTypes.function.isRequires
// };

export default Pagination;
