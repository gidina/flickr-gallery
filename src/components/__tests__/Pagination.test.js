
import React from "react";
import ReactTestUtils from "react-dom/test-utils";
import renderer from "react-test-renderer";
import Pagination from "../Pagination";

test("Renders as many photos as total/number of elements per pagee", () => {
    const numItemsPage = 8;
    const tree = renderer
        .create(<Pagination numItems={numItemsPage} onPageChange={() => {}} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test("First page if default active page", () => {
    const numItemsPage = 8;

    const onPageChangeMock = jest.fn();
    const tree = ReactTestUtils.renderIntoDocument(<Pagination numItems={numItemsPage} onPageChange={onPageChangeMock} />);

    let paginationElements = ReactTestUtils.scryRenderedDOMComponentsWithTag(tree, "li");
    paginationElements.shift();  // Removes the first li element (<<)

    const firstPage = paginationElements[0];

    expect(firstPage.className).toEqual("active");
    expect(onPageChangeMock).toHaveBeenNthCalledWith(1, 0, numItemsPage - 1);
});

test("When a page is clicked, is selected (active)", () => {
    const numItemsPage = 8;

    const onPageChangeMock = jest.fn();
    const tree = ReactTestUtils.renderIntoDocument(<Pagination numItems={numItemsPage} onPageChange={onPageChangeMock} />);
    let paginationElements = ReactTestUtils.scryRenderedDOMComponentsWithTag(tree, "li");
    paginationElements.shift();  // Removes the first li element (<<)
    paginationElements.pop();    // Removes the last li element (>>)

    const indexPage = 1;
    const secondPage = paginationElements[indexPage];

    expect(secondPage.className).toEqual("");

    ReactTestUtils.Simulate.click(secondPage, indexPage);

    expect(secondPage.className).toEqual("active");

    const startIndexItem = indexPage * numItemsPage;
    const endIndexItem = startIndexItem + (numItemsPage - 1);

    expect(onPageChangeMock).toHaveBeenNthCalledWith(2, startIndexItem, endIndexItem);
});

test("When back button is clicked, go back to the previous page", () => {
    const numItemsPage = 8;

    const onPageChangeMock = jest.fn();
    const tree = ReactTestUtils.renderIntoDocument(<Pagination numItems={numItemsPage} onPageChange={onPageChangeMock} />);
    let paginationElements = ReactTestUtils.scryRenderedDOMComponentsWithTag(tree, "li");

    const backLi = paginationElements[0]; // <<
    const indexPage = 1 + 1; // first element is back li element
    const firstPage = paginationElements[indexPage-1];
    const secondPage = paginationElements[indexPage];

    ReactTestUtils.Simulate.click(secondPage, indexPage);

    expect(secondPage.className).toEqual("active");

    ReactTestUtils.Simulate.click(backLi, indexPage - 1);

    expect(secondPage.className).toEqual("");
    expect(firstPage.className).toEqual("active");

    const startIndexItem = (indexPage - 1) * numItemsPage;
    const endIndexItem = startIndexItem + (numItemsPage - 1);

    expect(onPageChangeMock).toHaveBeenNthCalledWith(2, startIndexItem, endIndexItem);
});

test("When forward button is clicked, go to the next page", () => {
    const numItemsPage = 8;
    const defaultActivePage = 0;
    const onPageChangeMock = jest.fn();

    const tree = ReactTestUtils.renderIntoDocument(<Pagination numItems={numItemsPage} onPageChange={onPageChangeMock} />);
    let paginationElements = ReactTestUtils.scryRenderedDOMComponentsWithTag(tree, "li");

    const forwardLi = paginationElements[paginationElements.length - 1]; // >>
    const firstPage = paginationElements[0+1]; // first element is back li element
    const secondPage = paginationElements[1+1]; // first element is back li element

    expect(firstPage.className).toEqual("active");

    ReactTestUtils.Simulate.click(forwardLi);

    expect(firstPage.className).toEqual("");
    expect(secondPage.className).toEqual("active");

    const startIndexItem = (defaultActivePage + 1) * numItemsPage;
    const endIndexItem = startIndexItem + (numItemsPage - 1);

    expect(onPageChangeMock).toHaveBeenNthCalledWith(2, startIndexItem, endIndexItem);
});