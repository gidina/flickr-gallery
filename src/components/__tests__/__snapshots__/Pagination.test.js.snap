// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Es renderitzen un total de pàgines corresponent al nombre de fotos total entre el nombre d'elements per pàgina 1`] = `
<div
  className="center"
>
  <ul
    className="pagination"
  >
    <li
      onClick={[Function]}
    >
      «
    </li>
    <li
      className="active"
      onClick={[Function]}
    >
      1
    </li>
    <li
      className={null}
      onClick={[Function]}
    >
      2
    </li>
    <li
      onClick={[Function]}
    >
      »
    </li>
  </ul>
</div>
`;
